Library/OcConsoleLib/TextOutputBuiltin.c

```diff
@@ -620,6 +82,107 @@ STATIC EFI_CONSOLE_CONTROL_SCREEN_MODE      mConsoleMode = EfiConsoleControlScre
 #define MIN_SUPPORTED_CONSOLE_WIDTH   (80)
 #define MIN_SUPPORTED_CONSOLE_HEIGHT  (25)
 
+/**
+  @param[in]  ConsoleFont           Font to use.
+  @param[in]  Char                  Char for which to find information.
+  @param[out] Page                  Font page in which char (or fallback char) is found.
+  @param[out] GlyphIndex            Index to glyph for char (or fallback char).
+  @param[in]  UseFallback           Whether to use fallback char if char is not found.
+
+  @retval EFI_SUCCESS               Successfully returned font data for character.
+  @retval EFI_WARN_UNKNOWN_GLYPH    Returned font data for fallback character because requested
+                                    character was not in font.
+  @retval EFI_LOAD_ERROR            No valid font data returned. Will only occur in abnormal circumstances
+                                    (namely, broken font data), or when UseFallback is FALSE.
+**/
+STATIC
+EFI_STATUS
+GetConsoleFontCharInfo (
+  IN  OC_CONSOLE_FONT       *ConsoleFont,
+  IN  CHAR16                Char,
+  OUT OC_CONSOLE_FONT_PAGE  **Page,
+  OUT UINT8                 *GlyphIndex,
+  IN  BOOLEAN               UseFallback
+  )
+{
+  EFI_STATUS  Status;
+  UINT16      PageNumber;
+  UINT8       PageChar;
+  UINT16      PageIndex;
+
+  ASSERT (Page != NULL);
+  ASSERT (GlyphIndex != NULL);
+
+  PageNumber = (Char >> 7) & 0x1FF;
+  PageChar   = Char & 0x7F;
+
+  *GlyphIndex = 0;
+
+  Status = EFI_SUCCESS;
+  do {
+    if ((PageNumber >= ConsoleFont->PageMin) && (PageNumber < ConsoleFont->PageMax)) {
+      PageIndex = PageNumber - ConsoleFont->PageMin + 1;
+      if (ConsoleFont->PageOffsets != NULL) {
+        PageIndex = ConsoleFont->PageOffsets[PageIndex - 1];
+      }
+
+      if (PageIndex != 0) {
+        *Page = &ConsoleFont->Pages[PageIndex - 1];
+        if ((PageChar >= (*Page)->CharMin) && (PageChar < (*Page)->CharMax)) {
+          ASSERT ((*Page)->Glyphs != NULL);
+          if ((*Page)->GlyphOffsets == NULL) {
+            *GlyphIndex = PageChar - (*Page)->CharMin + 1;
+          } else {
+            *GlyphIndex = (*Page)->GlyphOffsets[PageChar - (*Page)->CharMin];
+          }
+        }
+      }
+    }
+
+    if (*GlyphIndex == 0) {
+      //
+      // Give up and render nothing if fallback char is not present.
+      //
+      ASSERT (Status == EFI_SUCCESS); ///< NB EFI_WARN_... are not errors
+      if (!UseFallback || (Status != EFI_SUCCESS)) {
+        return EFI_LOAD_ERROR;
+      }
+
+      Status = EFI_WARN_UNKNOWN_GLYPH;
+
+      //
+      // For default font, render all unknown chars in provided page(s) as space (following the explicit
+      // spaces in free console font used in XNU); in other pages and in user fonts use underscore.
```

Comment:
Not sure that we need the special case of rendering spaces in default font - unless exactly matching what e.g. Mac BIOS might render, in the (I guess) unlikely event of using these chars, really matters?

Reply:
Yeah, space is nicer because it does not immediately raise an issue that something is wrong. Maybe we want it for other fonts too really.

Reply:
Hi - Thank you for the review.

So, do we want any change to the current PR here?

We previously used space in the provided page of the provided font, and underscore outside it. My understanding of the use of space for non-present chars within the provided page of the provided font, in the previous code, was that it was because the original free font, as used in XNU, actually had blank char data at these code points - so not storing these zeroes, but rendering space when these chars were requested, was a minor storage optimisation.

However, I'm not sure there's any reason that anything would ever render glyphs from 0x0-0x1f, or 0x7f, which is what this applies to; and so, if this does happen, I think I'd prefer to see an underscore, and have a hint that's something's up, rather than to see a space and probably assume it was meant to be a space.

i.e. I think on balance I'd propose leaving it as is in the PR: we effectively keep the existing font as if it provided these characters, as blank glyphs, and other fonts provide just what they provide, and outside what is provided, we render underscore.

Reply:
Okay thanks for response by PM - will merge as is for now